fdb-record-layer-spatial/src/main/java/com/apple/foundationdb/record/spatial/geophile/GeophileSpatialFunctionKeyExpression.java

```diff
@@ -0,0 +1,214 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.spatial.geophile;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.metadata.Key;
+import com.apple.foundationdb.record.metadata.expressions.FunctionKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpression;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecord;
+import com.geophile.z.Space;
+import com.geophile.z.SpatialObject;
+import com.geophile.z.spatialobject.d2.Point;
+import com.google.protobuf.Message;
+import org.locationtech.jts.io.ParseException;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Spatial function key expressions.
+ */
+@API(API.Status.EXPERIMENTAL)
+public abstract class GeophileSpatialFunctionKeyExpression extends FunctionKeyExpression {
+    public static final Space SPACE_LAT_LON = GeophileSpatial.createLatLonSpace();
+
+    private final Space space;
+
+    protected GeophileSpatialFunctionKeyExpression(@Nonnull String name, @Nonnull KeyExpression arguments) {
+        super(name, arguments);
+        // TODO: How do we make this a part of the key expression? A naming convention?
+        this.space = SPACE_LAT_LON;
+    }
+
+    @Nonnull
+    public Space getSpace() {
+        return space;
+    }
+
+    @Nullable
+    protected abstract SpatialObject parseSpatialObject(@Nonnull Key.Evaluated arguments) throws ParseException;
+
+    @Nonnull
+    @Override
+    public <M extends Message> List<Key.Evaluated> evaluateFunction(@Nullable FDBRecord<M> record, @Nullable Message message, @Nonnull Key.Evaluated arguments) {
+        SpatialObject spatialObject;
+        try {
+            spatialObject = parseSpatialObject(arguments);
+        } catch (ParseException ex) {
+            throw new RecordCoreException(ex);
+        }
+        if (spatialObject == null) {
+            return Collections.singletonList(Key.Evaluated.NULL);
+        }
+        long[] zs = new long[spatialObject.maxZ()];
+        GeophileSpatial.shuffle(space, spatialObject, zs);
+        List<Key.Evaluated> result = new ArrayList<>(zs.length);
+        for (long z : zs) {
+            if (z == Space.Z_NULL) {
+                break;
+            }
+            result.add(Key.Evaluated.scalar(z));
+        }
+        return result;
+    }
+
+    @Override
+    public int getMinArguments() {
+        return 1;
+    }
+
+    @Override
+    public int getMaxArguments() {
+        return 2;
```

Comment:
The ones that parse text / byte strings take a second optional argument for whether to swap the x's and y's in the parsed thing when mapping to lat's and long's. The point one(s) overrides these methods. Do you think it would be clearer if they stayed abstract and everybody had to implement?

Reply:
Yeah, that's what I was suggesting, as I think each function individually expects either exactly 1 or exactly 2 arguments (I think?), so I think implementing them in each of them would avoid someone using, say, `GeoPointZ` but giving it only one argument.

Reply:
`GeoPointZ` takes exactly two and overrides to accomplish this.
The rest take one required and one optional argument and inherit.
Obviously it is confusing the way this is implemented now, so I'll change that.